rc/lsp.kak

```diff
@@ -228,18 +228,157 @@ define-command lsp-hover -docstring "Request hover info for the main cursor posi
+' "${kak_session}" "${kak_client}" "${kak_buffile}" "${kak_opt_filetype}" "${kak_timestamp}" ${kak_cursor_line} ${kak_cursor_column} "$1" "$2" | eval ${kak_opt_lsp_cmd} --request) > /dev/null 2>&1 < /dev/null & }
+}
+
+declare-option -hidden str lsp_symbol_kind_completion %{
+  symbol_kinds="
+  File Module Namespace Package Class Method Property Field Constructor Enum Interface Function
+  Variable Constant String Number Boolean Array Object Key Null EnumMember Struct Event Operator
+  TypeParameter
+  "
+  for symbol_kind in ${symbol_kinds}; do
+    printf '%s\n' "${symbol_kind}"
+  done
+}
+
+define-command lsp-goto-previous-symbol -params 0..1 -shell-script-candidates %opt{lsp_symbol_kind_completion} \
+    -docstring "lsp-goto-previous-symbol [<symbol-kind>]: Goto to previous document symbol of type <symbol-kind>. If <symbol-kind> is not given, it means _any_ symbol" %{
+    lsp-did-change-and-then "lsp-goto-previous-symbol-request %arg{@}"
+}
+
+define-command -hidden lsp-goto-previous-symbol-request -params 0..1 -shell-script-candidates %opt{lsp_symbol_kind_completion} \
+    -docstring "lsp-goto-previous-symbol-request [<symbol-kind>]: Goto to the previous symbol of type <symbol-kind> in the document. If <symbol-kind> is not given, it means _any_ symbol" %{
+    nop %sh{ (printf '
+session   = "%s"
+client    = "%s"
+buffile   = "%s"
+filetype  = "%s"
+version   = %d
+method    = "kak-lsp/next-or-previous-symbol"
+[params]
+position.line      = %d
+position.column    = %d
+symbolKind = "%s"
+searchNext = false
+' "${kak_session}" "${kak_client}" "${kak_buffile}" "${kak_opt_filetype}" "${kak_timestamp}" ${kak_cursor_line} ${kak_cursor_column} "$1" | eval ${kak_opt_lsp_cmd} --request) > /dev/null 2>&1 < /dev/null & }
+}
+
+define-command lsp-goto-next-symbol -params 0..1 -shell-script-candidates %opt{lsp_symbol_kind_completion} \
+    -docstring "lsp-goto-next-symbol [<symbol-kind>]: Goto to the next symbol of type <symbol-kind> in the document. If <symbol-kind> is not given, it means _any_ symbol" %{
+    lsp-did-change-and-then "lsp-goto-next-symbol-request %arg{@}"
+}
+
+define-command -hidden lsp-goto-next-symbol-request -params 0..1 -shell-script-candidates %opt{lsp_symbol_kind_completion} \
+    -docstring "lsp-goto-next-symbol-request [<symbol-kind>]: Goto to the next symbol of type <symbol-kind> in the document. If <symbol-kind> is not given, it means _any_ symbol" %{
+    nop %sh{ (printf '
+session   = "%s"
+client    = "%s"
+buffile   = "%s"
+filetype  = "%s"
+version   = %d
+method    = "kak-lsp/next-or-previous-symbol"
+[params]
+position.line      = %d
+position.column    = %d
+symbolKind = "%s"
+searchNext = true
+' "${kak_session}" "${kak_client}" "${kak_buffile}" "${kak_opt_filetype}" "${kak_timestamp}" ${kak_cursor_line} ${kak_cursor_column} "$1" | eval ${kak_opt_lsp_cmd} --request) > /dev/null 2>&1 < /dev/null & }
+}
+
+# Frequently used, provided for convenience
+define-command lsp-goto-next-function -params 0 \
```

Comment:
I'm not sure if it's worth to provide these convenience methods.
Without it, users will find the more powerful `lsp-goto-next-symbol` faster, and for bindings it doesn't matter.

Reply:
I actually like these conveniences and would request they remain for now. I'm OK if you want to remove them in the future. This feature is useful in language servers where you get symbols for variables and it quickly becomes tedious to goto them instead of functions which you're mainly interested in.

Of course, you could do `lsp-next-symbol Function` but that's a bit more verbose.

Reply:
Ok I guess it's useful to give the function/method variants more exposure since they are the 80% here.

> I'm OK if you want to remove them in the future.

Well I'm not gonna randomly remove features.
But we can decide later, of course.